crates/symbolicator-service/src/caches/versions.rs

```diff
@@ -8,13 +8,73 @@
 //!
 //! * increase the `current` version.
 //! * prepend the `current` version to the `fallbacks`.
+//! * add the `current` version to the `previous` versions.
 //! * it is also possible to skip a version, in case a broken deploy needed to
 //!   be reverted which left behind broken cache files.
 //!
 //! Some of the versioned caches are also tied to format versions defined in [`symbolic`].
 //! For those cases, there are static assertions that are a reminder to also bump the cache version.
 
-use crate::caching::CacheVersions;
+use std::fmt;
+
+/// How to format cache keys into file paths.
+#[derive(Clone, Copy, Debug)]
+pub enum CachePathFormat {
+    /// Format cache keys as `xx/xxxxxx/xxx…`.
+    V1,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct CacheVersion {
+    /// The version number.
+    pub number: u32,
+    /// The way in which cache keys should be formatted
+    /// into file paths for this version.
+    pub path_format: CachePathFormat,
+}
+
+impl CacheVersion {
+    /// Creates a new `CacheVersion` with the given number and path format.
+    pub const fn new(number: u32, path_format: CachePathFormat) -> Self {
+        Self {
+            number,
+            path_format,
+        }
+    }
+}
+
+impl fmt::Display for CacheVersion {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.number)
+    }
+}
+
+impl PartialEq for CacheVersion {
+    fn eq(&self, other: &Self) -> bool {
+        self.number == other.number
+    }
+}
+
+impl Eq for CacheVersion {}
+
+/// Cache Version Configuration used during cache lookup and generation.
+///
+/// The `current` version is tried first, and written during cache generation.
+/// The `fallback` versions are tried next, in first to last order. They are used only for cache
+/// lookups, but never for writing.
+///
+/// The version `0` is special in the sense that it is not used as part of the resulting cache
+/// file path, and generates the same paths as "legacy" unversioned cache files.
+#[derive(Clone, Debug)]
+pub struct CacheVersions {
+    /// The current cache version that is being looked up, and used for writing
+    pub current: CacheVersion,
+    /// A list of fallback cache versions that are being tried on lookup,
+    /// in descending order of priority.
+    pub fallbacks: &'static [CacheVersion],
+    /// A list of all previous cache versions.
+    pub previous: &'static [CacheVersion],
```

Comment:
Would clarify here if a version in `fallbacks` is *also* supposed to be in `previous` or not.

(looking through the rest of the changes, it seems to be they should be listed twice - which makes sense)